### PR TITLE
build_providers: add a build image setup facility

### PR DESCRIPTION
**Pull request by sergiusens**
Download the correct build image to build a base for the corresponding architecture.
Use a cached download if available and setup the original build disk image as the
backing store for the newly setup image.

LP: #1782779

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----
You test the UI by doing
```python
from snapcraft.internal.build_providers import images
images.setup(base="core18", snap_arch="amd64", size="4G", image_path="image.qcow2")
```

### CODE REVIEW

File: snapcraft/internal/build_providers/images.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                checksum="6d663a8fd5eddd916f4aef4fd06d0f7f4cf0bb191f170b8c84cd2adf297bc5c3",  # noqa: E501
+                algorithm="sha256",
+            )
+        )
+    else:
+        raise KeyError(base)
+
+
+def setup(*, base: str, snap_arch: str, size: str, image_path: str) -> None:
+    """Setup an instance for base and snap_arch on sized size image_path."""
```
Comment (kyrofa) [inline snapcraft/internal/build_providers/images.py]: I'm not parsing this description: "... on sized size image_path" ? Also, mind adding sphinx docs?

File: snapcraft/internal/build_providers/images.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    cached_file = image.get()
+
+    if os.path.dirname(image_path):
+        os.makedirs(os.path.dirname(image_path), exist_ok=True)
+    qemu_img_cmd = get_tool_path("qemu-img")
+    # qemu-img parameters:
+    # -q: quiet.
+    # -f: first image format.
+    # -b: backing file.
```
Comment (kyrofa) [inline snapcraft/internal/build_providers/images.py]: Thank you for this.

File: snapcraft/internal/build_providers/images.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        self._image_cache = FileCache(taxonomy="build-images-{}".format(self.base))
+
+    def _download_and_cache(self) -> str:
+        request = requests.get(self.url, stream=True, allow_redirects=True)
+        if not request.ok:
+            raise errors.BuildImageRequestError(
+                base=self.base, status_code=request.status_code
+            )
+        # First create the prefix as tempfile.TemporaryDirectory does not do that for you
+        os.makedirs(self._image_cache.file_cache, exist_ok=True)
```
Comment (kyrofa) [inline snapcraft/internal/build_providers/images.py]: \o/

File: snapcraft/internal/build_providers/images.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
```
Comment (kyrofa) [inline snapcraft/internal/build_providers/images.py]: Can you elaborate a little regarding the placement of this file? We have `build_providers/_qemu` and `build_providers/_multipass`, and now a `build_providers/images.py` that is very qemu-specific? What is the plan for multipass support with this interface?

File: snapcraft/internal/cache/_file.py [hunk 5]
```diff
@@ -26,12 +26,12 @@
 class FileCache(SnapcraftCache):
     """Generic file cache."""
 
-    def __init__(self):
+    def __init__(self, *, taxonomy: str = "files") -> None:
```
Comment (kyrofa) [inline snapcraft/internal/cache/_file.py]: Github decided to hide my comment, so I'll try again:

I don't know what to think when I see "taxonomy" as an argument. It's a field of study. From how it's used, perhaps "namespace" would make more sense?

Comment (sergiusens) [reply snapcraft/internal/cache/_file.py]: I was having troubles thinking of a word. Your suggestion is so obvious now!

File: snapcraft/internal/build_providers/images.py (hunk 4 above)
Comment (sergiusens) [reply snapcraft/internal/build_providers/images.py]: We will use these to launch the mutlipass images too.

File: snapcraft/internal/build_providers/images.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+
+def _get_build_images(base: str) -> Dict[str, _Image]:
+    if base == "core16":
+        return dict(
+            amd64=_Image(
+                base="core16",
+                snap_arch="amd64",
+                url="https://cloud-images.ubuntu.com/releases/16.04/release-20180703/ubuntu-16.04-server-cloudimg-amd64-disk1.img",  # noqa: E501
+                checksum="79549e87ddfc61b1cc8626a67ccc025cd7111d1af93ec28ea46ba6de70819f8c",  # noqa: E501
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/internal/build_providers/images.py]: Hm, yes, the notes. I am going on the fact that it is easier to remove this requirement in the future. Also, the fact that we are using this as a backing store means we need to have different entities for each project (related to pruning). I also had download issues :-)

Eventually we could keep this list online and marshal it from some json living on snapcore.

File: snapcraft/file_utils.py [hunk 7]
```diff
@@ -349,10 +349,16 @@ def get_tool_path(command_name):
...
         return path
+    elif shutil.which(command_name):
+        return shutil.which(command_name)
     else:
-        return command_name
+        fmt = (
+            "A tool snapcraft depends on could not be found: {command!r}.\n"
+            "Ensure the tool is installed and available, and try again."
+        )
+        raise RequiredCommandNotFound(fmt=fmt, command=command_name)
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/file_utils.py]: I will create a new one and add a TODO, your suggestion was my original idea, but the changes delve too deep into deprecated logic.

File: snapcraft/file_utils.py [hunk 8]
```diff
@@ -349,10 +349,16 @@ def get_tool_path(command_name):
             os.path.join(os.sep, "snap", "snapcraft", "current"), command_name
         )
 
-    if path:
+    if path is not None:
         return path
+    elif shutil.which(command_name):
+        return shutil.which(command_name)
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/file_utils.py]: assignment expressions will come soon :-P

[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/file_utils.py]: did is fixed in #2193

File: snapcraft/file_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/file_utils.py]: this is mitigated in #2193

File: snapcraft/internal/build_providers/images.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/internal/build_providers/images.py]: s/on/of/ :man_shrugging:

File: snapcraft/internal/build_providers/images.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply snapcraft/internal/build_providers/images.py]: Ah, okay.

**Pull request merged**